File: api/src/main/java/fi/ehin/resource/PriceResource.java
package fi.ehin.resource;

import fi.ehin.repository.PriceRepository;
import fi.ehin.service.PricesService;
import fi.ehin.utils.DateUtils;
import io.quarkus.logging.Log;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.QueryParam;
import jakarta.ws.rs.core.MediaType;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.OffsetTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.List;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.jboss.resteasy.reactive.RestResponse;

import static fi.ehin.utils.DateUtils.HELSINKI_ZONE;

@Path("/api")
public class PriceResource {

  private final PriceRepository priceRepository;
  private final PricesService pricesService;

  private static final String CACHE_VAR = "public";
  private static final String CACHE_LONG =
    "public, max-age=" + 60 * 60 * 168 + ", immutable";


  @ConfigProperty(name = "update-prices.password")
  String updatePricesPassword;

  public PriceResource(
    PriceRepository priceRepository,
    PricesService pricesService
  ) {
    this.priceRepository = priceRepository;
    this.pricesService = pricesService;
  }

  @GET
  @Produces(MediaType.APPLICATION_JSON)
  @Path("/prices/{date}")
  public RestResponse<List<PriceRepository.PriceHistoryEntry>> getPastPrices(
    @PathParam("date") final LocalDate date
  ) {
    Log.infof("Fetching prices for %s", date);

    final var dateWithTime = date.atTime(
      OffsetTime.of(
        0,
        0,
        0,
        0,
        HELSINKI_ZONE.getRules().getOffset(LocalDateTime.now())
      )
    );

    // The latest price is the day after's 0-1 prices. So plusDays 3 and for checking plusDays 2
    final var prices = priceRepository.getPrices(
      dateWithTime.minusDays(1),
      dateWithTime.plusDays(3)
    );
    String cacheString = CACHE_LONG; // Long cache if already have all prices
    String expiresValue = null;
    final var newestPriceDate = prices.getLast().deliveryStart().toLocalDate();
    if (!newestPriceDate.equals(date.plusDays(2))) {
      // Wait seconds until 11:57:30 but at least 60 seconds
      final var waitTime = DateUtils.secondsUntil12Utc() - 60*2+30;
      if(waitTime < 60) {
        Log.infof("Returning 60 seconds cache for prices on %s", date);
        cacheString = CACHE_VAR + ", max-age=60";
      } else {
        expiresValue = DateUtils.getGmtStringForCache();
        cacheString = CACHE_VAR;
        Log.infof("Returning expires %s for prices on %s", expiresValue, date);
      }
    }
    return RestResponse.ResponseBuilder.ok(prices)
      .header("Cache-Control", cacheString)
      .header("Expires", expiresValue)
      .build();
  }

  public record UpdatePricesResponse(boolean done) {}

  @GET
  @Produces(MediaType.APPLICATION_JSON)
  @Path("/update-prices")
  public RestResponse<UpdatePricesResponse> updatePrices(
    @QueryParam("p") final String password
  ) {
    if (!updatePricesPassword.equals(password)) {
      return RestResponse.notFound();
    }
    Log.infof("Updating prices for tomorrow %s", LocalDate.now().plusDays(1));
    final var prices = pricesService.GetTomorrowsPrices();
    if (prices == null) {
      return RestResponse.ok(new UpdatePricesResponse(false));
    }
    priceRepository.insertPrices(prices.multiAreaEntries());

    return RestResponse.ok(new UpdatePricesResponse(true));
  }

  @GET
  @Produces(MediaType.APPLICATION_JSON)
  @Path("/update-prices/{date}")
  public RestResponse<UpdatePricesResponse> updatePricesForDate(
    @PathParam("date") final LocalDate date,
    @QueryParam("p") final String password
  ) {
    if (!updatePricesPassword.equals(password)) {
      return RestResponse.notFound();
    }
    Log.infof("Updating prices for %s", date);
    final var prices = pricesService.GetPrices(date);
    priceRepository.insertPrices(prices.multiAreaEntries());

    return RestResponse.ok(new UpdatePricesResponse(true));
  }
}

File: api/internal/api/resource/price_resource.go
package resource

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/samlof/ehin/internal/db/repository"
	"github.com/samlof/ehin/internal/service"
	"github.com/samlof/ehin/internal/utils"
)

type PriceResource struct {
	priceRepository      repository.PriceRepository
	pricesService        *service.PricesService
	dateService          service.TimeProvider
	updatePricesPassword string
}

func NewPriceResource(
	priceRepository repository.PriceRepository,
	pricesService *service.PricesService,
	dateService service.TimeProvider,
	updatePricesPassword string,
) *PriceResource {
	return &PriceResource{
		priceRepository:      priceRepository,
		pricesService:        pricesService,
		dateService:          dateService,
		updatePricesPassword: updatePricesPassword,
	}
}

type UpdatePricesResponse struct {
	Done bool `json:"done"`
}

func (res *PriceResource) UpdatePrices(w http.ResponseWriter, r *http.Request) {
	password := r.URL.Query().Get("p")
	if res.updatePricesPassword == "" || res.updatePricesPassword != password {
		http.Error(w, "Not Found", http.StatusNotFound)
		return
	}

	slog.Info("Updating prices", "date", time.Now().AddDate(0, 0, 1).Format("2006-01-02"))
	prices, err := res.pricesService.GetTomorrowsPrices()
	if err != nil {
		slog.Error("Error fetching tomorrow's prices", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	if prices == nil {
		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(UpdatePricesResponse{Done: false}); err != nil {
			slog.Error("Error encoding response", "error", err)
		}
		return
	}

	entries := res.pricesService.ToPriceHistoryEntries(prices)
	_, err = res.priceRepository.InsertPrices(r.Context(), entries)
	if err != nil {
		slog.Error("Error inserting prices", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(UpdatePricesResponse{Done: true}); err != nil {
		slog.Error("Error encoding response", "error", err)
	}
}

func (res *PriceResource) UpdatePricesForDate(w http.ResponseWriter, r *http.Request) {
	password := r.URL.Query().Get("p")
	if res.updatePricesPassword == "" || res.updatePricesPassword != password {
		http.Error(w, "Not Found", http.StatusNotFound)
		return
	}

	dateStr := r.PathValue("date")
	date, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		http.Error(w, "Invalid date format. Use YYYY-MM-DD", http.StatusBadRequest)
		return
	}

	slog.Info("Updating prices", "date", dateStr)
	prices, err := res.pricesService.GetPrices(date)
	if err != nil {
		slog.Error("Error fetching prices", "date", dateStr, "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	if prices == nil {
		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(UpdatePricesResponse{Done: false}); err != nil {
			slog.Error("Error encoding response", "error", err)
		}
		return
	}

	entries := res.pricesService.ToPriceHistoryEntries(prices)
	_, err = res.priceRepository.InsertPrices(r.Context(), entries)
	if err != nil {
		slog.Error("Error inserting prices", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(UpdatePricesResponse{Done: true}); err != nil {
		slog.Error("Error encoding response", "error", err)
	}
}

func (res *PriceResource) GetPastPrices(w http.ResponseWriter, r *http.Request) {
	dateStr := r.PathValue("date")
	date, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		http.Error(w, "Invalid date format. Use YYYY-MM-DD", http.StatusBadRequest)
		return
	}

	slog.Info("Fetching prices", "date", dateStr)

	if res.priceRepository == nil {
		slog.Warn("Price repository not initialized")
		http.Error(w, "Database connection not available", http.StatusInternalServerError)
		return
	}

	helsinki, err := time.LoadLocation("Europe/Helsinki")
	if err != nil {
		slog.Error("Error loading Europe/Helsinki", "error", err)
		helsinki = time.UTC
	}

	// Helsinki 00:00:00 on the requested date
	dateWithTime := time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, helsinki)

	// Range matching Java implementation: date-1 to date+3
	from := dateWithTime.AddDate(0, 0, -1)
	to := dateWithTime.AddDate(0, 0, 3)

	prices, err := res.priceRepository.GetPrices(r.Context(), from, to)
	if err != nil {
		slog.Error("Error fetching prices from repository", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	cacheString := utils.CACHE_LONG
	var expiresValue string

	if len(prices) > 0 {
		// Java: newestPriceDate.isAfter(date) ? CACHE_LONG : (isAfter(11:57) ? CACHE_VAR+max-age=60 : EXPIRES+CACHE_VAR)
		lastPrice := prices[len(prices)-1]
		lastPriceDate := lastPrice.DeliveryStart.In(helsinki)
		lastPriceDateOnly := time.Date(lastPriceDate.Year(), lastPriceDate.Month(), lastPriceDate.Day(), 0, 0, 0, 0, helsinki)
		
		if !lastPriceDateOnly.After(dateWithTime) {
			pricesUpdateTime := time.Date(date.Year(), date.Month(), date.Day(), 11, 57, 0, 0, time.UTC)
			if res.dateService.Now().After(pricesUpdateTime) {
				cacheString = utils.CACHE_VAR + ", max-age=60"
			} else {
				expiresValue = utils.GetGmtStringForCache(pricesUpdateTime)
				cacheString = utils.CACHE_VAR
			}
		}
	} else {
		// No prices found, use same logic as if we don't have tomorrow's prices yet
		pricesUpdateTime := time.Date(date.Year(), date.Month(), date.Day(), 11, 57, 0, 0, time.UTC)
		if res.dateService.Now().After(pricesUpdateTime) {
			cacheString = utils.CACHE_VAR + ", max-age=60"
		} else {
			expiresValue = utils.GetGmtStringForCache(pricesUpdateTime)
			cacheString = utils.CACHE_VAR
		}
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set(utils.CACHE_CONTROL_HEADER, cacheString)
	if expiresValue != "" {
		w.Header().Set(utils.EXPIRES_HEADER, expiresValue)
	}

	if err := json.NewEncoder(w).Encode(prices); err != nil {
		slog.Error("Error encoding prices", "error", err)
	}
}
